pgx-pg-sys/src/submodules/datum.rs

```diff
@@ -0,0 +1,148 @@
+#[repr(C)]
+pub struct DatumBlob {
+    _data: [u8; 0],
+    _marker: core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
+}
+
+/// Datum is an abstract value that is effectively a union of all scalar types
+/// and all possible pointers in a Postgres context. That is, it is either
+/// "pass-by-value" (if the value fits into the platform's `uintptr_t`) or
+/// "pass-by-reference" (if it does not).
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Datum(*mut DatumBlob);
+
+impl Datum {
+    /// Assume the datum is a value and extract the bits from
+    /// the memory address, interpreting them as an integer.
+    pub fn value(self) -> usize {
+        self.0.addr()
+    }
+
+    /// Assume the datum is a pointer and treat it as void*.
+    pub fn to_void(self) -> *mut core::ffi::c_void {
+        self.0.cast()
+    }
+
+    /// True if the datum is equal to the null pointer.
+    pub fn is_null(self) -> bool {
+        self.0.is_null()
+    }
+
+    /// Assume the datum is a pointer and cast it to point to T.
+    pub fn ptr_cast<T>(self) -> *mut T {
+        self.0.cast()
+    }
+}
+
+impl From<usize> for Datum {
+    fn from(val: usize) -> Datum {
+        Datum(NonNull::<DatumBlob>::dangling().as_ptr().with_addr(val))
+    }
+}
+
+impl From<Datum> for usize {
+    fn from(val: Datum) -> usize {
+        val.0.addr()
+    }
+}
+
+impl From<isize> for Datum {
+    fn from(val: isize) -> Datum {
+        Datum::from(val as usize)
+    }
+}
+
+impl From<u8> for Datum {
+    fn from(val: u8) -> Datum {
+        Datum::from(usize::from(val))
+    }
+}
+
+impl From<u16> for Datum {
+    fn from(val: u16) -> Datum {
+        Datum::from(usize::from(val))
+    }
+}
+
+impl From<u32> for Datum {
+    fn from(val: u32) -> Datum {
+        Datum::from(val as usize)
+    }
+}
+
+impl From<u64> for Datum {
+    fn from(val: u64) -> Datum {
+        Datum::from(val as usize)
+    }
+}
+
+impl From<i8> for Datum {
+    fn from(val: i8) -> Datum {
+        Datum::from(isize::from(val))
+    }
+}
+
+impl From<i16> for Datum {
+    fn from(val: i16) -> Datum {
+        Datum::from(isize::from(val))
+    }
+}
+
+impl From<i32> for Datum {
+    fn from(val: i32) -> Datum {
+        Datum::from(val as usize)
+    }
+}
+
+impl From<i64> for Datum {
+    fn from(val: i64) -> Datum {
+        Datum::from(val as usize)
+    }
+}
+
+impl<T> From<*mut T> for Datum {
+    fn from(val: *mut T) -> Datum {
+        Datum(val.cast())
+    }
+}
+
+impl<T> From<*const T> for Datum {
+    fn from(val: *const T) -> Datum {
+        Datum(val as *mut _)
+    }
+}
+
+impl<T> PartialEq<*mut T> for Datum {
+    fn eq(&self, other: &*mut T) -> bool {
+        &self.0.cast() == other
+    }
+}
+
+impl<T> PartialEq<Datum> for *mut T {
+    fn eq(&self, other: &Datum) -> bool {
+        self == &other.0.cast()
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct NullableDatum {
```

Comment:
I didn't see this used anywhere.  I can intuit from the name/definition what it is, but what's your plan for it?

Reply:
It's one of the types that Postgres exposes through the bindgen. I wanted to bring both directly "into" our control, because it's important that NullableDatum actually uses the Datum type we define, otherwise it's too hard to use from user code, and I didn't want to leave it to chance such that it might use the wrong type as a field. If there are genuinely no other use sites in PGX **and** we expect that no one externally relies on this type, however, it would be fine to **not** define it on the condition we also block it from being exported by the bindgen.

Reply:
A few things...

- `NullableDatum` was introduced in Postgres 12.  It may not matter that we're now defining it for pg10/11 (as it wouldn't be used), but maybe we shouldn't pollute the namespace for them (also, we'll be dropping pg10 support in November of this year)
- There are some usages of it in `fcinfo.rs`.  Do they need to be double-checked, or did you and found them to be okay?
- Some doccomments on the type definition itself would be cool

And I'm sure you know, but Postgres' definition for it is below.  I don't see any discrepancies tho.

```c
/*
 * A NullableDatum is used in places where both a Datum and its nullness needs
 * to be stored. This can be more efficient than storing datums and nullness
 * in separate arrays, due to better spatial locality, even if more space may
 * be wasted due to padding.
 */
typedef struct NullableDatum
{
#define FIELDNO_NULLABLE_DATUM_DATUM 0
	Datum		value;
#define FIELDNO_NULLABLE_DATUM_ISNULL 1
	bool		isnull;
	/* due to alignment padding this could be used for flags for free */
} NullableDatum;
```

Reply:
Sure, we can make it feature dependent.

I was thinking it would eventually also be an obvious choice to do conversion impls for Datum and NullableDatum and it's "more obvious" if it's _right there_, so in a way it's there for "please remember to care about me some day, eventually this will matter".

Reply:
It'd be neat if our "nullable datum" could be `Option<Datum>`, but that ain't quite compatible, is it?

Reply:
Rust's ADT enums aren't compatible over FFI except in very specific "blessed" cases, but we could probably implement enough conversions and maybe with enough pixie dust it would feel almost transparent.

Reply:
Postgres' `NullableDatum` is really only used in the function call stuff.  Maybe it's also used elsewhere, but that's the only place **we** use it.  Since all that is abstracted away from the programmer, it doesn't matter.

I was just sayin'.  C sux

Reply:
Then probably we can manage to banish it from external use behind a layer of shims and such. I'll look into that as a followup.